dds_web/utils.py

```diff
@@ -749,3 +750,172 @@ def use_sto4(unit_object, project_object) -> bool:
+
+                        db.session.add(new_file)
+                        db.session.commit()
+                        files_added.append(new_file)
+                except (
+                    sqlalchemy.exc.IntegrityError,
+                    sqlalchemy.exc.OperationalError,
+                    sqlalchemy.exc.SQLAlchemyError,
+                ) as err:
+                    errors[file] = {"error": str(err)}
+                    db.session.rollback()
+    if errors:
+        flask.current_app.logger.error(f"Error in new_file_version: {errors}")
+
+    return files_added, errors
+
+
+def new_file_version(existing_file, new_info):
+    """
+    Create new version of a file.
+
+    Args:
+        existing_file (dds_web.models.File): The existing file to create a new version of.
+        new_info (dict): A dictionary containing information about the new version of the file.
+
+    Returns:
+        None
+    """
+    from dds_web import db
+    import dds_web.utils
+
+    # Get project
+    project = existing_file.project
+
+    # Get versions
+    current_file_version = models.Version.query.filter(
+        sqlalchemy.and_(
+            models.Version.active_file == sqlalchemy.func.binary(existing_file.id),
+            models.Version.time_deleted.is_(None),
+        )
+    ).all()
+
+    # If there is more than one version of the file which does not yet have a deletion timestamp, log a warning
+    if len(current_file_version) > 1:
+        flask.current_app.logger.warning(
+            "There is more than one version of the file "
+            "which does not yet have a deletion timestamp."
+        )
+
+    # Same timestamp for deleted and created new version
+    new_timestamp = dds_web.utils.current_time()
+
+    # Set the deletion timestamp for the latests version of the file
+    for version in current_file_version:
+        if version.time_deleted is None:
+            version.time_deleted = new_timestamp
+
+    # Update file info
+    existing_file.subpath = new_info["subpath"]
+    existing_file.size_original = new_info["size_raw"]
+    existing_file.size_stored = new_info["size_processed"]
+    existing_file.compressed = not new_info["compressed"]
+    existing_file.salt = new_info["salt"]
+    existing_file.public_key = new_info["public_key"]
+    existing_file.time_uploaded = new_timestamp
+    existing_file.checksum = new_info["checksum"]
+
+    # Create a new version of the file
+    new_version = models.Version(
+        size_stored=new_info["size_processed"],
+        time_uploaded=new_timestamp,
+        active_file=existing_file.id,
+        project_id=project,
+    )
+
+    # Update foreign keys and relationships
+    project.file_versions.append(new_version)
+    existing_file.versions.append(new_version)
+
+    # Add the new version to the database and commit the changes
+    db.session.add(new_version)
+    db.session.commit()
+
+    # Clean up information
+    del new_info
+    gc.collect()
```

Comment:
can I ask why is the Garbage collector manually called here? I am mostly curious

Reply:
Just added it because I wanted to make sure it was deleted, but also not entirely sure if it's needed here.

Reply:
I think the collector just frees space. Apparently, checking that an object still exists can be done counting the number of references it still has. But seems a bit tricky to use https://docs.python.org/3/library/sys.html#sys.getrefcount \\ https://towardsdatascience.com/understanding-reference-counting-in-python-3894b71b5611